File: python/api/tests/pitch_test.py
from python.api.classes.pitch import Pitch
import math

# since its python, everything should be in snake_case

def test_default_pitch():
    p = Pitch()
    assert p.swara == 0
    assert p.oct == 0
    assert p.raised == True
    assert p.fundamental == 261.63
    ratios = [
        1,
        [2 ** (1 / 12), 2 ** (2 / 12)],
        [2 ** (3 / 12), 2 ** (4 / 12)],
        [2 ** (5 / 12), 2 ** (6 / 12)],
        2 ** (7 / 12),
        [2 ** (8 / 12), 2 ** (9 / 12)],
        [2 ** (10 / 12), 2 ** (11 / 12)]
    ]
    assert p.ratios == ratios
    assert p.log_offset == 0.0

    # everything above here has been implemented
    assert p.frequency == 261.63
""" assert p.non_offset_frequency == 261.63
    logFreq = math.log2(p.frequency)
    assert p.non_offset_log_freq == logFreq
    assert p.log_freq == logFreq
    assert p.sargam_letter == 'S'
    assert p.octaved_sargam_letter == 'S'
    assert p.numbered_pitch == 0
    assert p.chroma == 0
    assert p.to_JSON() == { # this should still be camelCase
      "swara": 0,
      "raised": True,
      "oct": 0,
      "ratios": ratios,
      "fundamental": 261.63,
      "logOffset": 0,
    }

def test_swara_input():
    def sa_test(p: Pitch):
        sa_freq = 261.63
        sa_log_freq = math.log2(sa_freq)
        assert p.swara == 0
        assert math.isclose(p.frequency, sa_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, sa_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'S'
        assert p.raised == True
        assert p.chroma == 0



    def re_lowered_test(p: Pitch):
        re_freq = 277.19
        re_log_freq = math.log2(re_freq)
        assert p.swara == 1
        assert math.isclose(p.frequency, re_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, re_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'r'
        assert p.raised == False
        assert p.chroma == 1
    
    def re_raised_test(p: Pitch):
        re_freq = 293.67
        re_log_freq = math.log2(re_freq)
        assert p.swara == 1
        assert math.isclose(p.frequency, re_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, re_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'R'
        assert p.raised == True
        assert p.chroma == 2
    
    def ga_lowered_test(p: Pitch):
        ga_freq = 311.13
        ga_log_freq = math.log2(ga_freq)
        assert p.swara == 2
        assert math.isclose(p.frequency, ga_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, ga_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'g'
        assert p.raised == False
        assert p.chroma == 3

    def ga_raised_test(p: Pitch):
        ga_freq = 329.63
        ga_log_freq = math.log2(ga_freq)
        assert p.swara == 2
        assert math.isclose(p.frequency, ga_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, ga_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'G'
        assert p.raised == True
        assert p.chroma == 4
    
    def ma_lowered_test(p: Pitch):
        ma_freq = 349.23
        ma_log_freq = math.log2(ma_freq)
        assert p.swara == 3
        assert math.isclose(p.frequency, ma_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, ma_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'm'
        assert p.raised == False
        assert p.chroma == 5
    
    def ma_raised_test(p: Pitch):
        ma_freq = 370
        ma_log_freq = math.log2(ma_freq)
        assert p.swara == 3
        assert math.isclose(p.frequency, ma_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, ma_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'M'
        assert p.raised == True
        assert p.chroma == 6
    
    def pa_test(p: Pitch):
        pa_freq = 392
        pa_log_freq = math.log2(pa_freq)
        assert p.swara == 4
        assert math.isclose(p.frequency, pa_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, pa_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'P'
        assert p.raised == True
        assert p.chroma == 7
    
    def dha_lowered_test(p: Pitch):
        dha_freq = 415.31
        dha_log_freq = math.log2(dha_freq)
        assert p.swara == 5
        assert math.isclose(p.frequency, dha_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, dha_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'd'
        assert p.raised == False
        assert p.chroma == 8
    
    def dha_raised_test(p: Pitch):
        dha_freq = 440.01
        dha_log_freq = math.log2(dha_freq)
        assert p.swara == 5
        assert math.isclose(p.frequency, dha_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, dha_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'D'
        assert p.raised == True
        assert p.chroma == 9
    
    def ni_lowered_test(p: Pitch):
        ni_freq = 466.17
        ni_log_freq = math.log2(ni_freq)
        assert p.swara == 6
        assert math.isclose(p.frequency, ni_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, ni_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'n'
        assert p.raised == False
        assert p.chroma == 10
    
    def ni_raised_test(p: Pitch):
        ni_freq = 493.89
        ni_log_freq = math.log2(ni_freq)
        assert p.swara == 6
        assert math.isclose(p.frequency, ni_freq, abs_tol=0.01)
        assert math.isclose(p.log_freq, ni_log_freq, abs_tol=0.01)
        assert p.sargam_letter == 'N'
        assert p.raised == True
        assert p.chroma == 11

    sa_vars = ['Sa', 'sa', 'S', 's', 0]
    for sa in sa_vars:
        p = Pitch({ 'swara': sa })
        sa_test(p)
        p = Pitch({ 'swara': sa, 'raised': False })
        # // shouldn't matter if you pass it raised = false, will revert to
        # raised == true 
        sa_test(p)
    
    re_vars = ['Re', 're', 'R', 'r', 1]
    for re in re_vars:
        p = Pitch({ 'swara': re, 'raised': False })
        re_lowered_test(p)
        p = Pitch({ 'swara': re })
        re_raised_test(p)
    
    ga_vars = ['Ga', 'ga', 'G', 'g', 2]
    for ga in ga_vars:
        p = Pitch({ 'swara': ga, 'raised': False })
        ga_lowered_test(p)
        p = Pitch({ 'swara': ga })
        ga_raised_test(p)
    
    ma_vars = ['Ma', 'ma', 'M', 'm', 3]
    for ma in ma_vars:
        p = Pitch({ 'swara': ma, 'raised': False })
        ma_lowered_test(p)
        p = Pitch({ 'swara': ma })
        ma_raised_test(p)
    
    pa_vars = ['Pa', 'pa', 'P', 'p', 4]
    for pa in pa_vars:
        p = Pitch({ 'swara': pa })
        pa_test(p)
        p = Pitch({ 'swara': pa, 'raised': False })
        # // shouldn't matter if you pass it raised = false, will revert to
        # raised == true
        pa_test(p)
    
    dha_vars = ['Dha', 'dha', 'D', 'd', 5]
    for dha in dha_vars:
        p = Pitch({ 'swara': dha, 'raised': False })
        dha_lowered_test(p)
        p = Pitch({ 'swara': dha })
        dha_raised_test(p)
    
    ni_vars = ['Ni', 'ni', 'N', 'n', 6]
    for ni in ni_vars:
        p = Pitch({ 'swara': ni, 'raised': False })
        ni_lowered_test(p)
        p = Pitch({ 'swara': ni })
        ni_raised_test(p)

def test_octave_input():
    p = Pitch({ 'swara': 'sa', 'oct': -2 })
    saDown2 = 'S' + '\u0324'
    saDown1 = 'S' + '\u0323'
    saPlus1 = 'S' + '\u0307'
    saPlus2 = 'S' + '\u0308'
    assert p.oct == -2
    assert p.octaved_sargam_letter == saDown2
    p.set_oct(-1)
    assert p.oct == -1
    assert p.octaved_sargam_letter == saDown1
    p.set_oct(0)
    assert p.oct == 0
    assert p.octaved_sargam_letter == 'S'
    p.set_oct(1)
    assert p.oct == 1
    assert p.octaved_sargam_letter == saPlus1
    p.set_oct(2)
    assert p.oct == 2
    assert p.octaved_sargam_letter == saPlus2

def test_log_offset():
    offset = 0.1
    p = Pitch({ 'log_offset': offset })
    assert p.log_offset == offset
    sa_freq = 261.63
    sa_log_freq = math.log2(sa_freq)
    offset_sa_log_freq = sa_log_freq + offset
    offset_sa_freq = 2 ** offset_sa_log_freq
    assert math.isclose(p.frequency, offset_sa_freq, abs_tol=0.01)
    assert math.isclose(p.log_freq, offset_sa_log_freq, abs_tol=0.01)
    assert math.isclose(p.non_offset_frequency, sa_freq, abs_tol=0.01)

def test_numbered_pitch():
    p = Pitch({ 'swara': 5, 'oct': -2 })
    assert p.numbered_pitch == -15
    p = Pitch({ 'swara': 2, 'oct': 0 })
    assert p.numbered_pitch == 4
    p = Pitch({ 'swara': 3, 'oct': 1 })
    assert p.numbered_pitch == 17

 """

File: src/js/tests/pitch.test.ts
import { expect, test } from 'vitest';
import { Pitch } from '../classes';

test('defaultPitch', () => {
  const p = new Pitch();
  expect(p).toBeInstanceOf(Pitch);
  expect(p.swara).toEqual(0);
  expect(p.oct).toEqual(0);
  expect(p.raised).toEqual(true);
  expect(p.fundamental).toEqual(261.63);
  const ratios = [
    1, 
    [2 ** (1 / 12), 2 ** (2 / 12)],
    [2 ** (3 / 12), 2 ** (4 / 12)],
    [2 ** (5 / 12), 2 ** (6 / 12)],
    2 ** (7 / 12),
    [2 ** (8 / 12), 2 ** (9 / 12)],
    [2 ** (10 / 12), 2 ** (11 / 12)]
  ]
  expect(p.ratios).toEqual(ratios);
  expect(p.logOffset).toEqual(0);
  expect(p.frequency).toEqual(261.63);
  expect(p.nonOffsetFrequency).toEqual(261.63);
  const logFreq = Math.log2(261.63);
  expect(p.nonOffsetLogFreq).toEqual(logFreq);
  expect(p.logFreq).toEqual(logFreq);
  expect(p.sargamLetter).toEqual('S');
  expect(p.octavedSargamLetter).toEqual('S');
  expect(p.numberedPitch).toEqual(0);
  expect(p.chroma).toEqual(0);
  expect(p.toJSON()).toEqual({
    swara: 0,
    raised: true,
    oct: 0,
    ratios: ratios,
    fundamental: 261.63,
    logOffset: 0,
  })
});

test('swaraInput', () => {
  const saTest = (p: Pitch) => {
    const saFreq = 261.63;
    const saLogFreq = Math.log2(saFreq);
    expect(p.swara).toEqual(0);
    expect(p.frequency).toEqual(saFreq);
    expect(p.logFreq).toEqual(saLogFreq);
    expect(p.sargamLetter).toEqual('S');
    expect(p.raised).toEqual(true);
    expect(p.chroma).toEqual(0);
  };
  const reLoweredTest = (p: Pitch) => {
    const reFreq = 277.19;
    const reLogFreq = Math.log2(reFreq);
    expect (p.swara).toEqual(1);
    expect(p.frequency).toBeCloseTo(reFreq);
    expect(p.logFreq).toBeCloseTo(reLogFreq);
    expect(p.sargamLetter).toEqual('r');
    expect(p.raised).toEqual(false);
    expect(p.chroma).toEqual(1);
  }
  const reRaisedTest = (p: Pitch) => {
    const reFreq = 293.67;
    const reLogFreq = Math.log2(reFreq);
    expect (p.swara).toEqual(1);
    expect(p.frequency).toBeCloseTo(reFreq);
    expect(p.logFreq).toBeCloseTo(reLogFreq);
    expect(p.sargamLetter).toEqual('R');
    expect(p.raised).toEqual(true);
    expect(p.chroma).toEqual(2);
  }
  const gaLoweredTest = (p: Pitch) => {
    const gaFreq = 311.13;
    const gaLogFreq = Math.log2(gaFreq);
    expect (p.swara).toEqual(2);
    expect(p.frequency).toBeCloseTo(gaFreq);
    expect(p.logFreq).toBeCloseTo(gaLogFreq);
    expect(p.sargamLetter).toEqual('g');
    expect(p.raised).toEqual(false);
    expect(p.chroma).toEqual(3);
  }
  const gaRaisedTest = (p: Pitch) => {
    const gaFreq = 329.63;
    const gaLogFreq = Math.log2(gaFreq);
    expect (p.swara).toEqual(2);
    expect(p.frequency).toBeCloseTo(gaFreq);
    expect(p.logFreq).toBeCloseTo(gaLogFreq);
    expect(p.sargamLetter).toEqual('G');
    expect(p.raised).toEqual(true);
    expect(p.chroma).toEqual(4);
  }
  const maLoweredTest = (p: Pitch) => {
    const maFreq = 349.23;
    const maLogFreq = Math.log2(maFreq);
    expect (p.swara).toEqual(3);
    expect(p.frequency).toBeCloseTo(maFreq);
    expect(p.logFreq).toBeCloseTo(maLogFreq);
    expect(p.sargamLetter).toEqual('m');
    expect(p.raised).toEqual(false);
    expect(p.chroma).toEqual(5);
  }
  const maRaisedTest = (p: Pitch) => {
    const maFreq = 370;
    const maLogFreq = Math.log2(maFreq);
    expect (p.swara).toEqual(3);
    expect(p.frequency).toBeCloseTo(maFreq);
    expect(p.logFreq).toBeCloseTo(maLogFreq);
    expect(p.sargamLetter).toEqual('M');
    expect(p.raised).toEqual(true);
    expect(p.chroma).toEqual(6);
  }
  const paTest = (p: Pitch) => {
    const paFreq = 392;
    const paLogFreq = Math.log2(paFreq);
    expect (p.swara).toEqual(4);
    expect(p.frequency).toBeCloseTo(paFreq);
    expect(p.logFreq).toBeCloseTo(paLogFreq);
    expect(p.sargamLetter).toEqual('P');
    expect(p.raised).toEqual(true);
    expect(p.chroma).toEqual(7);

  }
  const dhaLoweredTest = (p: Pitch) => {
    const dhaFreq = 415.31;
    const dhaLogFreq = Math.log2(dhaFreq);
    expect (p.swara).toEqual(5);
    expect(p.frequency).toBeCloseTo(dhaFreq);
    expect(p.logFreq).toBeCloseTo(dhaLogFreq);
    expect(p.sargamLetter).toEqual('d');
    expect(p.raised).toEqual(false);
    expect(p.chroma).toEqual(8);
  }
  const dhaRaisedTest = (p: Pitch) => {
    const dhaFreq = 440.01;
    const dhaLogFreq = Math.log2(dhaFreq);
    expect (p.swara).toEqual(5);
    expect(p.frequency).toBeCloseTo(dhaFreq);
    expect(p.logFreq).toBeCloseTo(dhaLogFreq);
    expect(p.sargamLetter).toEqual('D');
    expect(p.raised).toEqual(true);
    expect(p.chroma).toEqual(9);
  }
  const niLoweredTest = (p: Pitch) => {
    const niFreq = 466.17;
    const niLogFreq = Math.log2(niFreq);
    expect (p.swara).toEqual(6);
    expect(p.frequency).toBeCloseTo(niFreq);
    expect(p.logFreq).toBeCloseTo(niLogFreq);
    expect(p.sargamLetter).toEqual('n');
    expect(p.raised).toEqual(false);
    expect(p.chroma).toEqual(10);
  }
  const niRaisedTest = (p: Pitch) => {
    const niFreq = 493.89;
    const niLogFreq = Math.log2(niFreq);
    expect (p.swara).toEqual(6);
    expect(p.frequency).toBeCloseTo(niFreq);
    expect(p.logFreq).toBeCloseTo(niLogFreq);
    expect(p.sargamLetter).toEqual('N');
    expect(p.raised).toEqual(true);
    expect(p.chroma).toEqual(11);
  }

  const saVars = ['Sa', 'sa', 's', 'S', 0]
  saVars.forEach((swara) => {
    let p = new Pitch({ swara: swara })
    saTest(p);
    p = new Pitch({ swara: swara, raised: false })
    saTest(p); // shouldn't matter if you pass it raised = false, will revert
    // to raised = true
  })

  const reVars = ['Re', 're', 'r', 'R', 1]
  reVars.forEach((swara) => {
    let p = new Pitch({ swara: swara })
    reRaisedTest(p);
    p = new Pitch({ swara: swara, raised: false })
    reLoweredTest(p);
  })

  const gaVars = ['Ga', 'ga', 'g', 'G', 2]
  gaVars.forEach((swara) => {
    let p = new Pitch({ swara: swara })
    gaRaisedTest(p);
    p = new Pitch({ swara: swara, raised: false })
    gaLoweredTest(p);
  })

  const maVars = ['Ma', 'ma', 'm', 'M', 3]
  maVars.forEach((swara) => {
    let p = new Pitch({ swara: swara })
    maRaisedTest(p);
    p = new Pitch({ swara: swara, raised: false })
    maLoweredTest(p);
  })

  const paVars = ['Pa', 'pa', 'p', 'P', 4]
  paVars.forEach((swara) => {
    let p = new Pitch({ swara: swara })
    paTest(p);
    p = new Pitch({ swara: swara, raised: false })
    paTest(p); // shouldn't matter if you pass it raised = false, will still be the same
  })

  const dhaVars = ['Dha', 'dha', 'd', 'D', 5]
  dhaVars.forEach((swara) => {
    let p = new Pitch({ swara: swara })
    dhaRaisedTest(p);
    p = new Pitch({ swara: swara, raised: false })
    dhaLoweredTest(p);
  })

  const niVars = ['Ni', 'ni', 'n', 'N', 6]
  niVars.forEach((swara) => {
    let p = new Pitch({ swara: swara })
    niRaisedTest(p);
    p = new Pitch({ swara: swara, raised: false })
    niLoweredTest(p);
  })
})

test('octaveInput', () => {
  // Only really expecting -2 - 2 to work
  let p = new Pitch({ swara: 'Sa', oct: -2 })
  const saDown2 = 'S' + '\u0324';
  const saDown1 = 'S' + '\u0323';
  const saPlus1 = 'S' + '\u0307';
  const saPlus2 = 'S' + '\u0308';
  expect(p.oct).toEqual(-2);
  expect(p.octavedSargamLetter).toEqual(saDown2);
  p.setOct(-1)
  expect(p.oct).toEqual(-1);
  expect(p.octavedSargamLetter).toEqual(saDown1);
  p.setOct(0)
  expect(p.oct).toEqual(0);
  expect(p.octavedSargamLetter).toEqual('S');
  p.setOct(1)
  expect(p.oct).toEqual(1);
  expect(p.octavedSargamLetter).toEqual(saPlus1);
  p.setOct(2)
  expect(p.oct).toEqual(2);
  expect(p.octavedSargamLetter).toEqual(saPlus2);

})

test('logOffset', () => {
  const offset = 0.1
  let p = new Pitch({ swara: 'Sa', logOffset: offset })
  expect(p.logOffset).toEqual(offset);
  const saFreq = 261.63;
  const saLogFreq = Math.log2(saFreq);
  const offsetSaLogFreq = saLogFreq + offset;
  const offsetSaFreq = Math.pow(2, offsetSaLogFreq);
  expect(p.frequency).toBeCloseTo(offsetSaFreq);
  expect(p.logFreq).toBeCloseTo(offsetSaLogFreq);
  expect(p.nonOffsetFrequency).toBeCloseTo(saFreq);
})

test('numberedPitch', () => {
  let p = new Pitch({ swara: 5, oct: -2})
  expect(p.numberedPitch).toEqual(-15);
  p = new Pitch({ swara: 2, oct: 0})
  expect(p.numberedPitch).toEqual(4);
  p = new Pitch({ swara: 3, raised: false, oct: 1 })
  expect(p.numberedPitch).toEqual(17);
})
